### DISCUSSION TITLE
How to Actually Run Criterion Tests?

### DISCUSSION THREAD

**Discussion by lordmichael95** [Snaipe/Criterion#516] [Q&A] [answered]
Hello All,

I am extremely new to unit testing and I am trying to use the Criterion Unit Test Framework for the first time in C (it seems perfect for my needs) . I have read these online documentation pages on Criterion (https://criterion.readthedocs.io/en/master/intro.html, https://criterion.readthedocs.io/en/master/setup.html, https://criterion.readthedocs.io/en/master/starter.html, https://criterion.readthedocs.io/en/master/assert.html, https://criterion.readthedocs.io/en/master/hooks.html, https://criterion.readthedocs.io/en/master/assert_old.html#old-assertions-ref). 

However, the one thing I did not see is how to actually run the Criterion test(s) once they have been written. Perhaps that should be added to the “Getting started” page (https://criterion.readthedocs.io/en/master/starter.html)? I would imagine that I need to pass some sort of pre-compiler option to Visual Studio Code or CMake? 

My current code is a small **main.c** and another file which contains all the Criterion testing code called **Michaels_tests_Criterion.c**.

**main.c**
```
#include <stdio.h>
#include <stdlib.h>
#include <signal.h>
#include <time.h>       //for clock() and CLOCKS_PER_SEC
#include "Michaels_test_Criterion.h"

#define TEST_SIGNAL_50 50       //PLACEHOLDER
#define TEST_SIGNAL_51 51       //PLACEHOLDER
#define TEST_SIGNAL_52 52       //PLACEHOLDER

sig_atomic_t signaled = 0;      //Debug testing only. Remove later.


/*
void sigHandler(int sig)
    The function is a signal handler.
    Parameters:
    int sig - The signal's number
    Returns:
    N/A
*/
void sigHandler(int sig)
{
    signaled = 1;

    //Switch statement to handle all signals
    switch(sig) {
        case TEST_SIGNAL_50:
            printf("Case 50 has been matched.\n");
            //LOG AND DO STUFF...
            break;
        case TEST_SIGNAL_51:
            printf("Case 51 has been matched.\n");
            //LOG AND DO STUFF...
            break;
        case TEST_SIGNAL_52:
            printf("Case 52 has been matched.\n");
            //LOG AND DO STUFF...
            break;
        case SIGINT:
            printf("SIGINT Case has been matched.\n");
            //LOG AND DO STUFF...
            break;
        default:    //Raised signal is not a match
            printf("Raised signal %d is not a match.\n", sig);
            //Probably just log an error and hope we can ignore without blowing up??
            break;
    }
}


/*

int timey(int trigger)
    This function is a timer which raises the signal after trigger milliseconds have elapsed
    Parameters:
    int trigger - The time, in milliseconds, in which this timer should trigger an alarm.
    Returns:
    int - This is the return result of raise(); zero if successful and a nonzero value if unsuccessful.
*/

int timey(int trigger){
    clock_t start = clock(), difference;
    int msec = 0;
    int raiseResult;       //Holds return value of raise(). raise() returns zero if successful and a nonzero value if unsuccessful.

    do{
        printf("This is a message\n");
        difference = clock() - start;                    //The difference between the start of the clock and "now".
        msec = difference * 1000 / CLOCKS_PER_SEC;       //CLOCKS_PER_SEC is defined in time.h. Clock ticks per second (system dependent)
    } while(msec < trigger);

    printf("Time taken %d seconds %d milliseconds\n", msec/1000, msec%1000);

    //raiseResult = raise(SIGINT);
    //raiseResult = raise(SIGUSR1);
    raiseResult = raise(TEST_SIGNAL_51);

    return raiseResult;
}

int main() {
    int raiseResult;
    void (*prev_handler)(int);

    printf("Hello, World!\n");

    //Register the signal handler
    //prev_handler = signal(SIGINT, sigHandler);
    prev_handler = signal(TEST_SIGNAL_51, sigHandler);
    //prev_handler = signal(SIGUSR1, sigHandler);

    //raiseResult = raise(SIGINT);
    //raiseResult = raise(TEST_SIGNAL_50);
    //raiseResult = raise(SIGUSR1);  

    raiseResult = timey(1);

    //debug
    printf("raiseResult =%d.\n",raiseResult );
    printf("signaled =%d.\n",signaled);

    //Test();       //I cannot seem to call my tests from main() without errors.

    return 0;

}
```

 **Michaels_tests_Criterion.c**
```
#include <stdio.h>
#include <stdlib.h>
#include <signal.h>
#include <criterion/criterion.h>            //For Criterion framework
#include "Michaels_test_Criterion.h"

int testValue;

void setup(void) {
    puts("setup: Runs before the test");

    testValue = 5;
}

void teardown(void) {
    puts("Runs after the test");
}

Test(suite1, test1, .init = setup, .fini = teardown) {
    cr_expect(testValue == 5);         //Should always pass
}

void setup2(void) {
    puts("setup2: Runs before the test");

    testValue = 4;
}


Test(suite1, test2, .init = setup2, .fini = teardown) {
    cr_expect(testValue == 5);         //Should always fail
}
```

These tests are mostly copy and pastes from the documentation. I am aware they do not test any specific part of my code in **main.c**. But I wanted to get these basic tests to run, before I start testing out signals from my signal handler. 

**Michaels_test_Criterion.h**
```
#ifndef MICHALES_TEST_CRITERION_H
#define MICHALES_TEST_CRITERION_H

//Function declarations
void setup(void);
void teardown(void);
void setup2(void);
//Do I need to declare the tests in here???


#endif /* MICHALES_TEST_CRITERION_H */
```

In short, how do I get my tests to actually run? Do I call them in **main.c** (I get errors when I do that)? I need to define the functions in **Michaels_test_Criterion.h**? Something else?

Any help would be appreciated! :)

Also, here is my CMake file if I am suppose to do this from CMake:
**CMakeLists.txt**
```
cmake_minimum_required(VERSION 3.20)
project(mainProject) #name this to whatever you'd like 

set(CMAKE_C_STANDARD 11)
#set(CMAKE_CXX_STANDARD 11)

#Begin testing...
#include(Criterion)
enable_testing()

add_executable(
    mainProject
    main.c
    Michaels_tests_Criterion.c
) 

target_link_libraries(
    mainProject
    criterion
)

#Maybe this is how??
add_test(
    NAME test1
    COMMAND $<TARGET_FILE:mainProject>
)
```

**Comment by Snaipe** [comment, marked as answer]
Hey there,

The main problem is that you're mixing your library code with your executable code. You seem to be linking criterion to your executable, which defines main, but your main overrides the default main that criterion provides.

Instead, you should introduce a third file that contains your code, and leave your main function in main.c.

You'd then have two executables: one for your main project, which links main.o and all of your library object files, and one for the tests, which links your library object files, your test object files, and libcriterion.

You can see this in action by compiling and linking directly one of the samples of this repo.

Alternatively, here's something I cooked up to show how this is supposed to be working: in the following directory, lib.c contains a small implementation of the fibonacci function, main.c is an executable that prints the nth fibonacci number, and test.c contains tests for the fibonacci function.

```console
$ cat <<'EOF' > lib.c
int fib(int n) {
    int a = 0, b = 1;
    for (int i = 0; i < n; i++) {
        int tmp = b;
        b = b + a;
        a = tmp;
    }
    return a;
}
EOF

$ cat <<'EOF' > lib.h
#ifndef LIB_H_
#define LIB_H_

int fib(int n);

#endif /* !LIB_H_ */
EOF

$ cat <<'EOF' > main.c
#include <stdio.h>
#include <stdlib.h>
#include "lib.h"

int main(int argc, char *argv[])
{
    printf("%d\n", fib(atoi(argv[1])));
    return 0;
}
EOF

$ cat <<'EOF' > test.c
#include <criterion/criterion.h>
#include <criterion/new/assert.h>
#include "lib.h"

Test(example, fib) {
    cr_assert(eq(int, fib(0), 0));
    cr_assert(eq(int, fib(1), 1));
    cr_assert(eq(int, fib(2), 1));
    cr_assert(eq(int, fib(3), 2));
    cr_assert(eq(int, fib(4), 3));
    cr_assert(eq(int, fib(5), 5));
    cr_assert(eq(int, fib(6), 8));
}

Test(example, bad) {
    cr_assert(eq(int, fib(0), 42));
}
EOF

$ # Building your application's executable
$ gcc -o fib main.c lib.c
$ ./fib 8
21

$ # Building the tests
$ gcc -o test test.c lib.c -lcriterion
$ ./test
[----] sample.c:16: Assertion Failed
[----]   eq(int, fib(0), 42): 
[----]     diff: [-0-]{+42+}
[FAIL] example::bad: (0.00s)
[====] Synthesis: Tested: 2 | Passing: 1 | Failing: 1 | Crashing: 0
```

**Reply by lordmichael95** [reply]
Thank you so much! :)
